### DISCUSSION TITLE
How to properly combine Bun Single-file executable + Prisma + Docker?

### DISCUSSION THREAD

**Discussion by akirarika** [prisma/prisma#22766] [Q&A] [answered]
### Question

I am using Prisma in Bun, and it works fine. However, I encountered several issues when trying to package it as a single file and create a Docker image. These problems mainly revolve around the need to embed Node.js in the image and the inability to locate the Prisma Engine (.x.so.node files) located in /node_modules/.prisma/client. I would like to ask if anyone has successfully done this and if there is a better solution at this stage?

**Comment by ludralph** [comment]
Hi @akirarika 👋 

Thank you for raising this question.

Some users have faced a similar issue before. A user mentioned that they faced a similar issue where `prisma generate` runs okay with Bun but fails to create a Prisma client when running with Docker. Their workaround was to generate the Prisma client locally with Bun and then ship it with the Docker image. They also changed the output of the generator to a different location to simplify the Dockerfile setup. See this [issue comment](https://github.com/prisma/prisma/issues/21241#issuecomment-1795144367). 

Another user got it working after installing Node.js on docker container. See [issue comment](https://github.com/prisma/prisma/issues/21241#issuecomment-1737212799).

Another user suggested using the `imbios/bun-node Docker image`, which contains both Node.js and Bun. This might help with the issue of embedding Node.js in the image. See [issue comment](https://github.com/prisma/prisma/issues/21241#issuecomment-1860605624)

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by akirarika** [reply]
Thank you very much for your help!😊

Before asking this question, I was able to successfully run Prisma + Bun + Docker without packaging, just as you said!

I actually encountered a problem when using Bun to package the entire project as a single-file. Prisma couldn't find the Prisma engine (.x.so.node file), even though I created a node_modules outside the single-file and placed the file in the original path. I am waiting to see if anyone has successfully combined these four (Bun + Single-file executable + Prisma + Docker) and shared their experience.

For now, I will continue to use Prisma without packaging.

Thank you again for your help! 😊

**Comment by medv** [comment]
Multiple prisma-related fixes have gone out in https://bun.sh/blog/bun-v1.0.30
Yet to test this myself, but we are in the same situation as you - push will come to shove in a couple of weeks and I will try to remember to come back here and update...

**Comment by pileskyd** [comment, marked as answer]
Good day!
I have encountered a similar problem, has anyone managed to package Bun + Prisma into a single executable file?

If I find any information, I will come back to the discussion and attach my solution

**Reply by akirarika** [reply]
I'm very upset, but after my research, it seems unlikely because the core part of Prisma is not written in TypeScript, and each system's binary files are different, which makes it difficult to combine Bun and Prisma. I ended up migrating to Drizzle.

**Reply by akirarika** [reply]
What I mean is, while bun and prisma can work together, it is quite difficult to bundle them into a single executable file, at least I have not found a solution, and then I gave up.

**Reply by wavim** [reply]
You could try to set `process.env.PRISMA_QUERY_ENGINE_LIBRARY` to
`path.join(import.meta.dirname, relativePathToNativeAddon)`.

**Comment by mlmetzner** [comment]
Hey I know this an old Discussion. But I ran in to similar issue. 

This my minimal image i came up with. It worked for me ( I am using an elysiajs ) 
Prisma is not bundled in the single executable, but this saves me around 600 Mb in imagesize

```yaml
FROM oven/bun AS build

WORKDIR /app

COPY package.json package.json
COPY bun.lockb bun.lockb

RUN bun install

COPY ./src ./src
COPY ./prisma ./prisma

ENV NODE_ENV=production
#Generate the Prisma Client for the distroless image
ENV PRISMA_CLI_BINARY_TARGETS="debian-openssl-3.0.x"

RUN bun prisma generate

RUN bun build \
    --compile \
    --minify-whitespace \
    --minify-syntax \
    --target bun \
    --outfile server \
    ./src/index.ts

# use the distroless cc images because prisma needs a cc libary
FROM gcr.io/distroless/cc

WORKDIR /app

COPY --from=build /app/server server
COPY --from=build /app/node_modules/.prisma /app/node_modules/.prisma

ENV NODE_ENV=production


CMD ["./server"]

EXPOSE 3000
```

**Reply by lacion** [reply]
this works great! thanks for that

**Comment by JoseDv1** [comment]
Prisma has relased a blog post where explains how to do it

https://www.prisma.io/docs/guides/bun#7-building-and-running-a-bun-executable
